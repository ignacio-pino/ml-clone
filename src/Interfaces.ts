export interface IProductListData {
  id: string;
  title: string;
  price: string;
  original_price?: string;
  currency_id: "UYU" | "USD";
  thumbnail: string;
  shipping: {
    free_shipping: boolean;
  };
}

export interface ISortDataList {
  sort: ISortData;
  available_sorts: ISortData[];
}

export interface ISortData {
  id: string;
  name: string;
}
export interface IProductCard {
  id: string;
  title: string;
  price: string;
  original_price?: string;
  currency_id: "UYU" | "USD";
  thumbnail: string;

  free_shipping: boolean;
}

export interface IFilterDataList {
  filters: IFilterList[];
  availableFilters: IFilterList[];
}
export interface IFilterList {
  id: string;
  name: string;
  type: string;
  values: IFilterListValues[];
}

export interface IFilterListValues {
  id: string;
  name: string;
  results?: number;
  path_from_root?: IPathFromRoot[];
}

export interface ICategoryData {
  query?: string;
  categoryInformation?: ICategoryInformation;
  totalResults: number;
  childrenCategories?: IPathFromRoot[];
}

export interface ICategoryInformation {
  id: string;
  name: string;
  path_from_root?: IPathFromRoot[];
}

export interface IPathFromRoot {
  id: string;
  name: string;
}

export interface IChildrenCategoryData {
  id: string;
  name: string;
  total_items_in_this_category: number;
}

export interface IHeaderCategoryData {
  categoryId: string;
  categoryName: string;
}
