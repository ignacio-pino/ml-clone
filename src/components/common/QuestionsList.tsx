import {
  Box,
  Button,
  HStack,
  Heading,
  Input,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalHeader,
  ModalOverlay,
  Text,
  VStack,
} from "@chakra-ui/react";
import { IQuestionsList } from "../../Interfaces";
import { useState } from "react";

const getFormattedDate = (dateData: string) => {
  const date = new Date(dateData);

  const day = String(date.getUTCDate()).padStart(2, "0");
  const month = String(date.getUTCMonth() + 1).padStart(2, "0"); // January is 0!
  const year = date.getUTCFullYear();

  return `${day}/${month}/${year}`;
};

export default function QuestionsList({
  questionsList,
}: {
  questionsList: IQuestionsList[];
}) {
  let shortenedQuestionsList: IQuestionsList[] = [];

  if (questionsList.length > 5) {
    shortenedQuestionsList = questionsList.slice(0, 5);
  }

  const [isModalOpen, setIsModalOpen] = useState(false);

  const questionListInformation = (list: IQuestionsList[]) =>
    list.map(({ id, status, text, answer }) => (
      <Box key={id} alignItems={"flex-start"} my={"1rem"}>
        <Text>{text}</Text>
        {status === "ANSWERED" && (
          <HStack pl={"1rem"}>
            <Text color={"gray.500"} fontSize={"l"} alignSelf={"flex-start"}>
              L
            </Text>
            <Text color={"gray.500"} mt={1}>
              {answer.text}{" "}
              <Text
                fontSize={"sm"}
                color={"gray.500"}
                as={"span"}
                whiteSpace={"nowrap"}
              >
                {getFormattedDate(answer.date_created)}
              </Text>
            </Text>
          </HStack>
        )}
      </Box>
    ));

  return (
    <>
      <VStack alignItems={"flex-start"} gap={7}>
        <Text>Preguntas y respuestas</Text>
        <HStack flexWrap={"wrap"}>
          <Input placeholder="Escribe tu pregunta" w={"md"} />
          <Button backgroundColor={"purple.400"}>Preguntar</Button>
        </HStack>
        <Box>
          {questionsList.length === 0 && (
            <Text color={"gray.500"}>
              {" "}
              Todavía no hay preguntas. ¡Sé el primero en preguntar!
            </Text>
          )}
          {questionsList && shortenedQuestionsList.length > 0
            ? questionListInformation(shortenedQuestionsList)
            : questionListInformation(questionsList)}
          {shortenedQuestionsList.length > 0 && (
            <Button
              variant={"link"}
              color={"purple.400"}
              onClick={() => setIsModalOpen(true)}
            >
              Mostrar todas las preguntas
            </Button>
          )}
        </Box>
      </VStack>

      <Modal
        isOpen={isModalOpen}
        onClose={() => setIsModalOpen(false)}
        size={"xl"}
        scrollBehavior="inside"
      >
        <ModalOverlay />
        <ModalContent>
          <ModalHeader>
            <Heading>Preguntas y respuestas</Heading>
            <ModalCloseButton />
          </ModalHeader>
          <ModalBody>{questionListInformation(questionsList)}</ModalBody>
        </ModalContent>
      </Modal>
    </>
  );
}
