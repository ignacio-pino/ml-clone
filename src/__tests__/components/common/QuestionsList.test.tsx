import { render, screen, fireEvent } from "@testing-library/react";
import QuestionsList from "../../../components/common/QuestionsList";
import { IQuestionsList } from "../../../Interfaces";

const sampleQuestions: IQuestionsList[] = [
  {
    id: "q1",
    status: "ANSWERED",
    text: "What is the product made of?",
    answer: {
      text: "The product is made of wood.",
      date_created: "2023-01-01T12:00:00Z",
    },
  },
  {
    id: "q2",
    status: "UNANSWERED",
    text: "Is the product available in other colors?",
    answer: { text: "", date_created: "" },
  },
  {
    id: "q3",
    status: "ANSWERED",
    text: "How much does the product weigh?",
    answer: {
      text: "The product weighs 2kg.",
      date_created: "2023-02-05T12:00:00Z",
    },
  },
  {
    id: "q4",
    status: "UNANSWERED",
    text: "Can the product be customized?",
    answer: { text: "", date_created: "" },
  },
  {
    id: "q5",
    status: "ANSWERED",
    text: "How long is the warranty for the product?",
    answer: {
      text: "The product has a 1-year warranty.",
      date_created: "2023-03-01T10:00:00Z",
    },
  },
  {
    id: "q6",
    status: "ANSWERED",
    text: "Is the product waterproof?",
    answer: {
      text: "Yes, the product is waterproof.",
      date_created: "2023-03-20T09:00:00Z",
    },
  },
];

describe("QuestionsList Component", () => {
  it("should match snapshot", () => {
    const container = render(
      <QuestionsList questionsList={sampleQuestions} />
    ).container;
    expect(container).toMatchSnapshot();
  });

  it("displays provided questions and answers", () => {
    render(<QuestionsList questionsList={sampleQuestions} />);

    const questionElement1 = screen.getByText("What is the product made of?");
    1;
    const answerElement1 = screen.getByText("The product is made of wood.");
    const questionElement3 = screen.getByText(
      "How much does the product weigh?"
    );
    const answerElement3 = screen.getByText("The product weighs 2kg.");
    const questionElement5 = screen.getByText(
      "How long is the warranty for the product?"
    );
    const answerElement5 = screen.getByText(
      "The product has a 1-year warranty."
    );

    expect(questionElement1).toBeInTheDocument();
    expect(answerElement1).toBeInTheDocument();
    expect(questionElement3).toBeInTheDocument();
    expect(answerElement3).toBeInTheDocument();
    expect(questionElement5).toBeInTheDocument();
    expect(answerElement5).toBeInTheDocument();
  });

  it("displays a message if there are no questions", () => {
    render(<QuestionsList questionsList={[]} />);

    const noQuestionsMessage = screen.getByText(
      "Todavía no hay preguntas. ¡Sé el primero en preguntar!"
    );
    expect(noQuestionsMessage).toBeInTheDocument();
  });

  it("shows only first 5 questions by default if there are more than 5 questions", () => {
    const extendedQuestions = [...sampleQuestions, ...sampleQuestions];

    render(<QuestionsList questionsList={extendedQuestions} />);

    const showAllQuestionsButton = screen.getByText(
      "Mostrar todas las preguntas"
    );
    expect(showAllQuestionsButton).toBeInTheDocument();
  });

  it("opens a modal when 'Mostrar todas las preguntas' button is clicked", () => {
    const extendedQuestions = [...sampleQuestions, ...sampleQuestions];

    render(<QuestionsList questionsList={extendedQuestions} />);

    const showAllQuestionsButton = screen.getByText(
      "Mostrar todas las preguntas"
    );
    fireEvent.click(showAllQuestionsButton);

    const modal = screen.getByRole("dialog");
    expect(modal).toBeInTheDocument();
  });

  it("correctly formats the date of the answer", () => {
    render(<QuestionsList questionsList={sampleQuestions} />);

    const formattedDate1 = screen.getByText("01/01/2023");
    const formattedDate3 = screen.getByText("05/02/2023");
    const formattedDate5 = screen.getByText("01/03/2023");

    expect(formattedDate1).toBeInTheDocument();
    expect(formattedDate3).toBeInTheDocument();
    expect(formattedDate5).toBeInTheDocument();
  });
});
