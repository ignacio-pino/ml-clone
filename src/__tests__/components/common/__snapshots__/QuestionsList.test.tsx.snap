// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuestionsList Component > should match snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-f5fnvy"
  >
    <p
      class="chakra-text css-0"
    >
      Preguntas y respuestas
    </p>
    <div
      class="chakra-stack css-zztjb4"
    >
      <input
        class="chakra-input css-1smfaui"
        placeholder="Escribe tu pregunta"
      />
      <button
        class="chakra-button css-1s8cab6"
        type="button"
      >
        Preguntar
      </button>
    </div>
    <div
      class="css-0"
    >
      <div
        class="css-1caknck"
      >
        <p
          class="chakra-text css-0"
        >
          What is the product made of?
        </p>
        <div
          class="chakra-stack css-1pkewcz"
        >
          <p
            class="chakra-text css-thg5cc"
          >
            L
          </p>
          <p
            class="chakra-text css-77frxq"
          >
            The product is made of wood.
             
            <span
              class="chakra-text css-131f7pn"
            >
              01/01/2023
            </span>
          </p>
        </div>
      </div>
      <div
        class="css-1caknck"
      >
        <p
          class="chakra-text css-0"
        >
          Is the product available in other colors?
        </p>
      </div>
      <div
        class="css-1caknck"
      >
        <p
          class="chakra-text css-0"
        >
          How much does the product weigh?
        </p>
        <div
          class="chakra-stack css-1pkewcz"
        >
          <p
            class="chakra-text css-thg5cc"
          >
            L
          </p>
          <p
            class="chakra-text css-77frxq"
          >
            The product weighs 2kg.
             
            <span
              class="chakra-text css-131f7pn"
            >
              05/02/2023
            </span>
          </p>
        </div>
      </div>
      <div
        class="css-1caknck"
      >
        <p
          class="chakra-text css-0"
        >
          Can the product be customized?
        </p>
      </div>
      <div
        class="css-1caknck"
      >
        <p
          class="chakra-text css-0"
        >
          How long is the warranty for the product?
        </p>
        <div
          class="chakra-stack css-1pkewcz"
        >
          <p
            class="chakra-text css-thg5cc"
          >
            L
          </p>
          <p
            class="chakra-text css-77frxq"
          >
            The product has a 1-year warranty.
             
            <span
              class="chakra-text css-131f7pn"
            >
              01/03/2023
            </span>
          </p>
        </div>
      </div>
      <button
        class="chakra-button css-18d007l"
        type="button"
      >
        Mostrar todas las preguntas
      </button>
    </div>
  </div>
</div>
`;
